File: 2022/day2/python/part2..py
def calculate_score(filename: str) -> int:
    with open(filename, 'r') as f:
        opponent_moves, results = zip(
            *[line.strip().split(" ") for line in f.readlines()])

    draw = {'A': 'R', 'B': 'P', 'C': 'S'}
    player_wins = {'A': 'P', 'B': 'S', 'C': 'R'}
    player_losses = {'A': 'S', 'B': 'R', 'C': 'P'}
    move_points = {'R': 1, 'P': 2, 'S': 3}

    total_points = 0

    for opponent_move, result in zip(opponent_moves, results):
        match result:
            case 'X':
                total_points += move_points[player_losses[opponent_move]]
            case 'Y':
                total_points += 3 + move_points[draw[opponent_move]]
            case 'Z':
                total_points += 6 + move_points[player_wins[opponent_move]]

    return total_points


def main():
    solution = calculate_score('./input.txt')
    print(solution)


main()

File: 2022/AoCJava/src/Day2/Part2.java
package Day2;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Hashtable;

public class Part2 {
    public static void main(String[] args) throws IOException {
        BufferedReader input = new BufferedReader(new FileReader("inputs/day2.txt"));

        Hashtable<String, String> draw = new Hashtable<>();
        draw.put("A", "R");
        draw.put("B", "P");
        draw.put("C", "S");

        Hashtable<String, String> playerLosses = new Hashtable<>();
        playerLosses.put("A", "S");
        playerLosses.put("B", "R");
        playerLosses.put("C", "P");

        Hashtable<String, String> playerWins = new Hashtable<>();
        playerWins.put("A", "P");
        playerWins.put("B", "S");
        playerWins.put("C", "R");

        Hashtable<String, String> movePoints = new Hashtable<>();
        movePoints.put("R", "1");
        movePoints.put("P", "2");
        movePoints.put("S", "3");

        int totalPoints = 0;

        String line;

        while ((line = input.readLine()) != null) {
            line = line.trim();
            String[] moves = line.split(" ");

            switch (moves[1]) {
                case "X" -> totalPoints += Integer.parseInt(movePoints.get(playerLosses.get(moves[0])));
                case "Y" -> totalPoints += 3 + Integer.parseInt(movePoints.get(draw.get(moves[0])));
                case "Z" -> totalPoints += 6 + Integer.parseInt(movePoints.get(playerWins.get(moves[0])));
            }
        }

        System.out.println(totalPoints);
    }
}
